Support pre-build hook with wails dev & wails build

### Is your feature request related to a problem? Please describe.

The problem is that my team already has built our front-end assets, and we just need to include it in our desktop app (https://github.com/ipfs/ipfs-desktop).

You can see in the README of https://github.com/SgtPooki/ipfs-desktop-wails, that we don't need a front-end command to run at all, but we do need a command to run to *get* our prebuilt frontend (see https://github.com/SgtPooki/ipfs-desktop-wails#run-a-local-version).

Since we need to get these front-end assets between `wails dev` and `wails generate module`, we get errors unless we run an explicit command prior to running `wails`. I believe it would be wise for wails to offer a solution for the multitude of usecases a preBuildHook could support.

### Describe the solution you'd like

`wails.json` already has `postBuildHooks`, I would like to propose `preBuildHooks`. Essentially an agnostic version of "frontend:*" though not necessarily a replacement.

For my issue, I would only need to run `./get-webui.sh` in the preBuildHook:

`wails.json`
```json
...
    "preBuildHooks": {
        "*/*": "./get-webui.sh"
    }
...
```

### Describe alternatives you've considered

The only two solutions I could think of that allow us to fully bundle our webui in the app (obtaining via runtime is not a viable solution) are:
1. require build process to run our `get-webui.sh` shell script prior to running wails command
2. move .get-webui.sh to `./frontend/get-webui.sh` and set `"frontend:build": "bash ./get-webui.sh",` in `wails.json`


### Solution 1
This works, and is okay, but definitely not ideal, requires build cmd to be:
```bash
./get-webui.sh && wails build # or wails dev
```
at least on the first run. 

### Solution 2

This is currently what I'm using as it's the desired DX I want. See https://github.com/SgtPooki/ipfs-desktop-wails/commit/fab24911d05a16ac0852414ad82086149b24404c

but allows us to have a single canonical wails build command: 
```bash
wails build # or wails dev
```

#### Some issues with this solution that others may run into

##### With no `./frontend` folder:

```bash
╰─ ✔ ❯ wails dev -v 2
Wails CLI v2.0.0-beta.38


Executing: go mod tidy -compat=1.17
Executing: wails generate module


ERROR:
main.go:10:12: pattern frontend: no matching files found
```

##### with existing `./frontend/.gitkeep`

```bash
╰─ ✔ ❯ wails dev -v 2
Wails CLI v2.0.0-beta.38


Executing: go mod tidy -compat=1.17
Executing: wails generate module


ERROR:
main.go:10:12: pattern frontend: cannot embed directory frontend: contains no embeddable files

exit status 1

If Wails is useful to you or your company, please consider sponsoring the project:
https://github.com/sponsors/leaanthony


exit status 1
```

##### With empty `./frontend/index.html`

This allows things to work, but I don't like the idea that wails could be modifying/hooking into a file that will be overwritten. 

### Additional context

_No response_

--- a/v2/internal/project/project.go
+++ b/v2/internal/project/project.go
@@ -51,15 +51,16 @@ type Project struct {
 	// RunNonNativeBuildHooks will run build hooks though they are defined for a GOOS which is not equal to the host os
 	RunNonNativeBuildHooks bool `json:"runNonNativeBuildHooks"`
 
-	// Post build hooks for different targets, the hooks are executed in the following order
-	// Key: GOOS/GOARCH - Executed at build level after a build of the specific platform and arch
-	// Key: GOOS/*      - Executed at build level after a build of the specific platform
-	// Key: */*         - Executed at build level after a build
+	// Build hooks for different targets, the hooks are executed in the following order
+	// Key: GOOS/GOARCH - Executed at build level before/after a build of the specific platform and arch
+	// Key: GOOS/*      - Executed at build level before/after a build of the specific platform
+	// Key: */*         - Executed at build level before/after a build
 	// The following keys are not yet supported.
-	// Key: GOOS        - Executed at platform level after all builds of the specific platform
-	// Key: *           - Executed at platform level after all builds of a platform
-	// Key: [empty]     - Executed at global level after all builds of all platforms
+	// Key: GOOS        - Executed at platform level before/after all builds of the specific platform
+	// Key: *           - Executed at platform level before/after all builds of a platform
+	// Key: [empty]     - Executed at global level before/after all builds of all platforms
 	PostBuildHooks map[string]string `json:"postBuildHooks"`
+	PreBuildHooks  map[string]string `json:"preBuildHooks"`
 
 	// The application author
 	Author Author

--- a/v2/pkg/commands/build/build.go
+++ b/v2/pkg/commands/build/build.go
@@ -115,6 +115,16 @@ func Build(options *Options) (string, error) {
 	// Initialise Builder
 	builder.SetProjectData(projectData)
 
+	hookArgs := map[string]string{
+		"${platform}": options.Platform + "/" + options.Arch,
+	}
+
+	for _, hook := range []string{options.Platform + "/" + options.Arch, options.Platform + "/*", "*/*"} {
+		if err := execPreBuildHook(outputLogger, options, hook, hookArgs); err != nil {
+			return "", err
+		}
+	}
+
 	if !options.IgnoreFrontend || options.ForceBuild {
 		err = builder.BuildFrontend(outputLogger)
 		if err != nil {
@@ -215,10 +225,7 @@ func Build(options *Options) (string, error) {
 	}
 
 	compileBinary := options.CompiledBinary
-	hookArgs := map[string]string{
-		"${platform}": options.Platform + "/" + options.Arch,
-		"${bin}":      compileBinary,
-	}
+	hookArgs["${bin}"] = compileBinary
 
 	for _, hook := range []string{options.Platform + "/" + options.Arch, options.Platform + "/*", "*/*"} {
 		if err := execPostBuildHook(outputLogger, options, hook, hookArgs); err != nil {
@@ -227,6 +234,55 @@ func Build(options *Options) (string, error) {
 	}
 
 	return compileBinary, nil
+}
+
+func execPreBuildHook(outputLogger *clilogger.CLILogger, options *Options, hookIdentifier string, argReplacements map[string]string) error {
+	preBuildHook := options.ProjectData.PreBuildHooks[hookIdentifier]
+	if preBuildHook == "" {
+		return nil
+	}
+
+	if !options.ProjectData.RunNonNativeBuildHooks {
+		if hookIdentifier == "" {
+			// That's the global hook
+		} else {
+			platformOfHook := strings.Split(hookIdentifier, "/")[0]
+			if platformOfHook == "*" {
+				// That's OK, we don't have a specific platform of the hook
+			} else if platformOfHook == runtime.GOOS {
+				// The hook is for host platform
+			} else {
+				// Skip a hook which is not native
+				outputLogger.Println("  - Non native build hook '%s': Skipping.", hookIdentifier)
+				return nil
+			}
+		}
+	}
+
+	outputLogger.Print("  - Executing pre build hook '%s': ", hookIdentifier)
+	args := strings.Split(preBuildHook, " ")
+	for i, arg := range args {
+		newArg := argReplacements[arg]
+		if newArg == "" {
+			continue
+		}
+		args[i] = newArg
+	}
+
+	if options.Verbosity == VERBOSE {
+		outputLogger.Println("%s", strings.Join(args, " "))
+	}
+
+	stdout, stderr, err := shell.RunCommand(options.BuildDirectory, args[0], args[1:]...)
+	if options.Verbosity == VERBOSE {
+		println(stdout)
+	}
+	if err != nil {
+		return fmt.Errorf("%s - %s", err.Error(), stderr)
+	}
+	outputLogger.Println("Done.")
+
+	return nil
 }
 
 func execPostBuildHook(outputLogger *clilogger.CLILogger, options *Options, hookIdentifier string, argReplacements map[string]string) error {
@@ -266,7 +322,10 @@ func execPostBuildHook(outputLogger *clilogger.CLILogger, options *Options, hook
 		outputLogger.Println("%s", strings.Join(args, " "))
 	}
 
-	_, stderr, err := shell.RunCommand(options.BuildDirectory, args[0], args[1:]...)
+	stdout, stderr, err := shell.RunCommand(options.BuildDirectory, args[0], args[1:]...)
+	if options.Verbosity == VERBOSE {
+		println(stdout)
+	}
 	if err != nil {
 		return fmt.Errorf("%s - %s", err.Error(), stderr)
 	}

--- a/v2/pkg/commands/build/build.go
+++ b/v2/pkg/commands/build/build.go
@@ -242,6 +242,20 @@ func execPreBuildHook(outputLogger *clilogger.CLILogger, options *Options, hookI
 		return nil
 	}
 
+	return executeBuildHook(outputLogger, options, hookIdentifier, argReplacements, preBuildHook, "pre")
+}
+
+func execPostBuildHook(outputLogger *clilogger.CLILogger, options *Options, hookIdentifier string, argReplacements map[string]string) error {
+	postBuildHook := options.ProjectData.PostBuildHooks[hookIdentifier]
+	if postBuildHook == "" {
+		return nil
+	}
+
+	return executeBuildHook(outputLogger, options, hookIdentifier, argReplacements, postBuildHook, "post")
+
+}
+
+func executeBuildHook(outputLogger *clilogger.CLILogger, options *Options, hookIdentifier string, argReplacements map[string]string, buildHook string, hookName string) error {
 	if !options.ProjectData.RunNonNativeBuildHooks {
 		if hookIdentifier == "" {
 			// That's the global hook
@@ -259,57 +273,8 @@ func execPreBuildHook(outputLogger *clilogger.CLILogger, options *Options, hookI
 		}
 	}
 
-	outputLogger.Print("  - Executing pre build hook '%s': ", hookIdentifier)
-	args := strings.Split(preBuildHook, " ")
-	for i, arg := range args {
-		newArg := argReplacements[arg]
-		if newArg == "" {
-			continue
-		}
-		args[i] = newArg
-	}
-
-	if options.Verbosity == VERBOSE {
-		outputLogger.Println("%s", strings.Join(args, " "))
-	}
-
-	stdout, stderr, err := shell.RunCommand(options.BuildDirectory, args[0], args[1:]...)
-	if options.Verbosity == VERBOSE {
-		println(stdout)
-	}
-	if err != nil {
-		return fmt.Errorf("%s - %s", err.Error(), stderr)
-	}
-	outputLogger.Println("Done.")
-
-	return nil
-}
-
-func execPostBuildHook(outputLogger *clilogger.CLILogger, options *Options, hookIdentifier string, argReplacements map[string]string) error {
-	postBuildHook := options.ProjectData.PostBuildHooks[hookIdentifier]
-	if postBuildHook == "" {
-		return nil
-	}
-
-	if !options.ProjectData.RunNonNativeBuildHooks {
-		if hookIdentifier == "" {
-			// That's the global hook
-		} else {
-			platformOfHook := strings.Split(hookIdentifier, "/")[0]
-			if platformOfHook == "*" {
-				// Thats OK, we don't have a specific platform of the hook
-			} else if platformOfHook == runtime.GOOS {
-				// The hook is for host platform
-			} else {
-				// Skip a hook which is not native
-				outputLogger.Println("  - Non native build hook '%s': Skipping.", hookIdentifier)
-				return nil
-			}
-		}
-	}
-
-	outputLogger.Print("  - Executing post build hook '%s': ", hookIdentifier)
-	args := strings.Split(postBuildHook, " ")
+	outputLogger.Print("  - Executing %s build hook '%s': ", hookName, hookIdentifier)
+	args := strings.Split(buildHook, " ")
 	for i, arg := range args {
 		newArg := argReplacements[arg]
 		if newArg == "" {

--- a/website/docs/reference/project-config.mdx
+++ b/website/docs/reference/project-config.mdx
@@ -23,11 +23,16 @@ The project config resides in the `wails.json` file in the project directory. Th
 	"devServer": "[Address to bind the wails dev sever to. Default: localhost:34115]",
 	"appargs": "[Arguments passed to the application in shell style when in dev mode]",
 	"runNonNativeBuildHooks": false, // Defines if build hooks should be run though they are defined for an OS other than the host OS.
-	"postBuildHooks": {
-		"GOOS/GOARCH": "[The command that will be executed after a build of the specified GOOS/GOARCH: ${platform} is replaced with the "GOOS/GOARCH" and ${bin} with the path to the compiled binary. The "GOOS/GOARCH" hook is executed before the "GOOS/*" and "*/*" hook.]",
-		"GOOS/*": "[The command that will be executed after a build of the specified GOOS: ${platform} is replaced with the "GOOS/GOARCH" and ${bin} with the path to the compiled binary. The "GOOS/*" hook is executed before the "*/*" hook.]",
-		"*/*": "[The command that will be executed after every build: ${platform} is replaced with the "GOOS/GOARCH" and ${bin} with the path to the compiled binary.]"
+	"preBuildHooks": {
+		"GOOS/GOARCH": "[The command that will be executed before a build of the specified GOOS/GOARCH: ${platform} is replaced with the "GOOS/GOARCH". The "GOOS/GOARCH" hook is executed before the "GOOS/*" and "*/*" hook.]",
+		"GOOS/*": "[The command that will be executed before a build of the specified GOOS: ${platform} is replaced with the "GOOS/GOARCH". The "GOOS/*" hook is executed before the "*/*" hook.]",
+		"*/*": "[The command that will be executed before every build: ${platform} is replaced with the "GOOS/GOARCH".]"
 	},
+    "postBuildHooks": {
+        "GOOS/GOARCH": "[The command that will be executed after a build of the specified GOOS/GOARCH: ${platform} is replaced with the "GOOS/GOARCH" and ${bin} with the path to the compiled binary. The "GOOS/GOARCH" hook is executed before the "GOOS/*" and "*/*" hook.]",
+        "GOOS/*": "[The command that will be executed after a build of the specified GOOS: ${platform} is replaced with the "GOOS/GOARCH" and ${bin} with the path to the compiled binary. The "GOOS/*" hook is executed before the "*/*" hook.]",
+        "*/*": "[The command that will be executed after every build: ${platform} is replaced with the "GOOS/GOARCH" and ${bin} with the path to the compiled binary.]"
+    },
 	"info": { // Data used to populate manifests and version info.
 		"companyName": "[The company name. Default: [The project name]]",
 		"productName": "[The product name. Default: [The project name]]",

